File: python/leetcode.236.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    fa = dict()

    def find(v):
        if not v in fa:
            fa[v] = v
        if fa[v] == v:
            return v
        else:
            return find(fa[v])

    def merge(a, b):
        aa = find(a)
        bb = find(b)
        if aa != bb:
            fa[aa] = bb

    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        
N = [1,2,3,4,5]
N = [1,2,None,3]
N1 = TreeNode(0)
build_tree(N, N1, 0)

sln = Solution()

print sln.lowestCommonAncestor(N1, N1.left.left, N1).val
print sln.lowestCommonAncestor(N1, N1.left.left, N1.right).val
print sln.lowestCommonAncestor(N1, N1.left.right, N1.right).val

# N = [-1,0,3,-2,4,None,None,8]
# N1 = TreeNode(0)
# build_tree(N, N1, 0)
# print sln.lowestCommonAncestor(N1, N1.left, N1.left.left.left).val # 0

# N = [3,5,1,6,2,0,8,None,None,7,4]
# N1 = TreeNode(0)
# build_tree(N, N1, 0)
# print sln.lowestCommonAncestor(N1, N1.left, N1.left.right.right).val # 5

# N = [3,5,1,6,2,0,8,None,None,7,4]
# N1 = TreeNode(0)
# build_tree(N, N1, 0)
# print sln.lowestCommonAncestor(N1, N1.left, N1.right).val # 3

File: cpp/leetcode.236.cpp
#include <iostream> 
#include <fstream>
#include <vector>
#include <queue>
#include <cstring>
#include <string>
#include <algorithm>
#include <stack>
#include <functional>
#include "stdlib.h" 
#include "time.h"
#include <set>
#include <map>
#include <numeric>
#include <cctype>
#include <cmath>
#include <sstream>

#define INF 0x3f3f3f3f
#define MOD 100000007
#define MAXN 510
using namespace std;
#define LL long long
#define ULL unsigned long long
#define LD long double
typedef pair<int, int> pii;


struct TreeNode {     
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

std::map<TreeNode *, TreeNode *> fa;

TreeNode * find(TreeNode * node) {
    if (fa.find(node) == fa.end()) {
        fa[node] = node;
    }
    while (fa[node] != node) {
        node = fa[node];
    }
    return node;
}

TreeNode * merge(TreeNode * a, TreeNode * b) {
    TreeNode * ffa = find(a);
    TreeNode * ffb = find(b);
    fa[ffb] = ffa;
    return ffa;
}

class Solution {
public:
    TreeNode* p = nullptr; 
    TreeNode* q = nullptr; 
    TreeNode* ans = nullptr;
    bool vp = false, vq = false;
    void dfs(TreeNode * cur) {
        if (cur->left)
        {
            dfs(cur->left);
            merge(cur, cur->left);
        }
        if (cur->right)
        {
            dfs(cur->right);
            merge(cur, cur->right);
        }
        if (cur == p)
        {
            vp = true;
            if (vq)
            {
                this->ans = find(q);
            }
        }
        if (cur == q)
        {
            vq = true;
            if (vq)
            {
                this->ans = find(p);
            }
        }
    }
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        fa.clear();
        this->p = p;
        this->q = q;
        ans = nullptr;
        vp = false, vq = false;
        dfs(root);
        return this->ans;
    }
};

int main() {
    std::vector<TreeNode *> t = { new TreeNode(999), new TreeNode(2), new TreeNode(3) };
    t[0]->left = t[1];
    t[0]->right = t[2];
    Solution sln = Solution();
    TreeNode * ans = sln.lowestCommonAncestor(t[0], t[1], t[2]);
    printf("%d\n", ans->val);

    t = { new TreeNode(999), new TreeNode(2), new TreeNode(3) };
    t[0]->left = t[1];
    t[1]->right = t[2];
    ans = sln.lowestCommonAncestor(t[0], t[0], t[2]);
    printf("%d\n", ans->val);

    system("pause");
}
